### DISCUSSION TITLE
Usage of formulas within mice and output of predictorMatrix

### DISCUSSION THREAD

**Discussion by tinneuro** [amices/mice#573] [Q&A] [answered]
Hi everyone,

I am performing multiple imputation for a single covariate in my dataset where I am specifying the formula to be used in mice. I am using logistic regression as method. 

The formula looks like this in principle:
```r
formula_list <- list(
  y~ x1+x2+ x3+ x4
  )
formula_list <- name.formulas(formula_list)
```
The dataset I put into the mice function has more columns/covariates (.i.e continuing `x5`-`x10`) which I need for the analysis after the imputation. Of course I want that only `x1`- `x4` are used in the imputation as specified by the formula. 
After I apply 
```r
mice(data,formulas = formula_list)
```
and looking at the obtained predictorMatrix I see that there is a 1 for all variables `x1` up to `x10`. How can I be sure that mice really uses the formula and not all variables?

Thanks!

**Comment by stefvanbuuren** [comment, marked as answer]
Good question. Here's a way to do it.

``` r
library(mice, warn.conflicts = FALSE)
f <- make.formulas(nhanes)
f$bmi <- bmi ~ 1
imp <- mice(nhanes, formulas = f, method = "norm.predict", print = FALSE)
imp$formulas
#> $age
#> age ~ bmi + hyp + chl
#> <environment: 0x10a3c3bf8>
#> 
#> $bmi
#> bmi ~ 1
#> 
#> $hyp
#> hyp ~ age + bmi + chl
#> <environment: 0x10a3c3bf8>
#> 
#> $chl
#> chl ~ age + bmi + hyp
#> <environment: 0x10a3c3bf8>
imp$imp$bmi
#>          1       2       3       4       5
#> 1  26.5625 26.5625 26.5625 26.5625 26.5625
#> 3  26.5625 26.5625 26.5625 26.5625 26.5625
#> 4  26.5625 26.5625 26.5625 26.5625 26.5625
#> 6  26.5625 26.5625 26.5625 26.5625 26.5625
#> 10 26.5625 26.5625 26.5625 26.5625 26.5625
#> 11 26.5625 26.5625 26.5625 26.5625 26.5625
#> 12 26.5625 26.5625 26.5625 26.5625 26.5625
#> 16 26.5625 26.5625 26.5625 26.5625 26.5625
#> 21 26.5625 26.5625 26.5625 26.5625 26.5625
imp$predictorMatrix
#>     age bmi hyp chl
#> age   0   1   1   1
#> bmi   1   0   1   1
#> hyp   1   1   0   1
#> chl   1   1   1   0
```

<sup>Created on 2023-07-21 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

What can learn from the example:

1. Use the convenience function `make.formulas()` to generate a valid `formulas` argument. It yields the default main effects formula for every variable;
2. Change the formula(s), e.g. by including `poly()`, `s()`, `I()` or other terms. In the example, I have specified to impute `bmi` by the intercept only; 
3. Run `mice` with the `formulas` argument. I have specified `meth = "norm.predict"` to show the effect of the intercept-only imputation model;
4. The `imp$formulas` components shows the imputation models actually used for imputation. `formulas` take precedence over `predictorMatrix` specification;
5. The `imp$imp$bmi` component shows that this specification just imputes the mean of the observed data, conform to the `bmi ~ 1` specification;
6. The `imp$predictorMatrix` is not updated with the `formulas` specification, so ignore this aspect of the return value.

Hope this helps using formulas.

**Reply by tinneuro** [reply]
Thanks for the fast and detailed answer! I appreciate it a lot.

I still have a follow up comment/question:
I now used the make.formulas() as you describe and it works. The version I posted before also ran without error.
The reason why I did it in the way of my first post:
 basically I have 11 variables in my dataset which I want to keep for computations later on. Only one of them has missing values and thiss ones I want to impute using a formula. If I use the make.formulas() then I get 11 equations for each variable but I only need one. I know that when running mice that then anyways only the one is imputed that has missings and the rest is left as it is but I find it a bit scary/annoying to have the other formulas there as well even though they should not be used.

**Reply by stefvanbuuren** [reply]
Simple. You can drop any formula entries you don't like. For example,

```{r}
imp <- mice(nhanes, formulas = list(bmi = bmi ~ 1 + chl), 
            method = "norm.predict", print = FALSE)
```

imputes only `bmi`. The downside is that you get `NA`s for rows where `chl` is missing (in your data that is not a problem). 

Note also that the `predictorMatrix` entry is not updated. Formulas work without the need to list all variables.

**Reply by tinneuro** [reply]
Thanks!
